jira/src/main/kotlin/gropius/sync/jira/JiraSync.kt

```diff
@@ -0,0 +1,368 @@
+
+/**
+ * Jira main sync class
+ * @param jiraDataService the data service for Jira
+ * @param helper the json helper
+ * @param cursorResourceWalkerDataService the data service for the cursor resource walker
+ * @param imsConfigManager the config manager for the IMS
+ * @param collectedSyncInfo the collected sync info
+ * @param loadBalancedDataFetcher the load balanced data fetcher
+ * @param issueDataService the data service for issues
+ */
+@Component
+final class JiraSync(
+    val jiraDataService: JiraDataService,
+    val helper: JsonHelper,
+    val cursorResourceWalkerDataService: CursorResourceWalkerDataService,
+    val imsConfigManager: IMSConfigManager,
+    collectedSyncInfo: CollectedSyncInfo,
+    val loadBalancedDataFetcher: LoadBalancedDataFetcher = LoadBalancedDataFetcher(),
+    val issueDataService: IssueDataService
+) : AbstractSync(collectedSyncInfo) {
+
+    private val client = HttpClient() {
+        expectSuccess = true
+        install(Logging)
+        install(ContentNegotiation) {
+            json(Json {
+                ignoreUnknownKeys = true
+            }, contentType = ContentType.parse("application/json; charset=utf-8"))
+        }
+    }
+
+    /**
+     * Logger used to print notifications
+     */
+    private val logger = LoggerFactory.getLogger(JiraSync::class.java)
+
+    override fun syncDataService(): SyncDataService {
+        return jiraDataService
+    }
+
+    override suspend fun findTemplates(): Set<IMSTemplate> {
+        return imsConfigManager.findTemplates()
+    }
+
+    override suspend fun isOutgoingEnabled(imsProject: IMSProject): Boolean {
+        val imsProjectConfig = IMSProjectConfig(helper, imsProject)
+        return imsProjectConfig.enableOutgoing
+    }
+
+    override suspend fun isOutgoingLabelsEnabled(imsProject: IMSProject): Boolean {
+        val imsProjectConfig = IMSProjectConfig(helper, imsProject)
+        return imsProjectConfig.enableOutgoingLabels
+    }
+
+    override suspend fun isOutgoingCommentsEnabled(imsProject: IMSProject): Boolean {
+        val imsProjectConfig = IMSProjectConfig(helper, imsProject)
+        return imsProjectConfig.enableOutgoingComments
+    }
+
+    override suspend fun isOutgoingTitleChangedEnabled(imsProject: IMSProject): Boolean {
+        val imsProjectConfig = IMSProjectConfig(helper, imsProject)
+        return imsProjectConfig.enableOutgoingTitleChanges
+    }
+
+    override suspend fun isOutgoingAssignmentsEnabled(imsProject: IMSProject): Boolean {
+        val imsProjectConfig = IMSProjectConfig(helper, imsProject)
+        return imsProjectConfig.enableOutgoingAssignments
+    }
+
+    override suspend fun isOutgoingStatesEnabled(imsProject: IMSProject): Boolean {
+        val imsProjectConfig = IMSProjectConfig(helper, imsProject)
+        return imsProjectConfig.enableOutgoingState
+    }
+
+    @OptIn(ExperimentalEncodingApi::class)
+    override suspend fun fetchData(imsProjects: List<IMSProject>) {
```

Comment:
this one is a bit long

Reply:
f8d2e54ecba22b78bcdb691d0a687b9a66e1c65e